reth: init at 1.3.12

mitchmindtree opened the pull request:
This commit adds `reth`, an Ethereum execution client by Paradigm written in Rust.

This package is adapted from the [nix-community/ethereum.nix package][1] with a version update, darwin support, and some slight modifications to better suit the nixpkgs conventions.

Closes #249233

[1]: https://github.com/nix-community/ethereum.nix/blob/main/packages/clients/execution/reth/default.nix

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

N/A

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mitchmindtree committed:
reth: init at 1.3.12

This commit adds `reth`, an Ethereum execution client by Paradigm
written in Rust.

This package is adapted from the [nix-community/ethereum.nix package][1]
with a version update, darwin support, and some slight modifications to
better suit the nixpkgs conventions.

Closes #249233

[1]: https://github.com/nix-community/ethereum.nix/blob/main/pkgs/by-name/re/reth/default.nix

mitchmindtree commented:
Hmmm looks like oborg-eval isn't happy about something, but it's not immediately obvious to me reading the description of looking at the trace:

```
This PR does not cleanly list package outputs after merging.
```
:thinking: 

The trace appears to point to some sort of issue related to `import-cargo-lock.nix` and `source` not being available?

```
       at /ofborg/checkout/1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-1/pkgs/build-support/rust/import-cargo-lock.nix:45:10:

           44|     if lockFile != null
           45|     then builtins.readFile lockFile
             |          ^
           46|     else args.lockFileContents;

       error: path '/nix/store/pbm5jyz0x5rg8l4pwsmrj8vk80q0hxn4-source.drv' is not valid
```

mitchmindtree commented:
Ahhh after taking a closer look at some other rust packages, I noticed most were including the `Cargo.lock` locally rather than using the `"${src}/Cargo.lock"`. I've updated the PR to follow this convention and it seems to have solved the issue mentioned above.

sambacha commented:
why not add this to the ethereum.nix flake - @aldoborrero

mitchmindtree commented:
> why not add this to the ethereum.nix flake

Already replied on TG but will do here for prosperity:

`reth` is already at [`ethereum.nix`](https://github.com/nix-community/ethereum.nix), which is very useful!

The main motivation for adding it to `nixpkgs` is that I can see `reth` getting a lot of usage, and having it as a part of the default nix binary cache would be a huge UX win for nix+reth users.

While flakes *kind* of allow specifying custom caches, the UX is still far from smooth (see discussion [here](https://discourse.nixos.org/t/ux-problems-with-flakes-and-custom-caches-substituters/23691)).

FliegendeWurst commented on pkgs/by-name/re/reth/package.nix:
```diff
@@ -0,0 +1,60 @@
+    rustPlatform.bindgenHook
+  ];
+
+  buildInputs = lib.optionals stdenv.isDarwin [
+    darwin.apple_sdk.frameworks.Security
+  ];
```
I think this is no longer needed.

FliegendeWurst commented on pkgs/by-name/re/reth/package.nix:
```diff
@@ -0,0 +1,60 @@
+    extraArgs = [ "--version" "unstable" ];
+  };
+
+  meta = with lib; {
+    description = "A modular Ethereum execution client in Rust by Paradigm";
+    homepage = "https://github.com/paradigmxyz/reth";
+    license = with licenses; [ mit asl20 ];
+    mainProgram = "reth";
+    maintainers = with maintainers; [ mitchmindtree ];
```
```suggestion
  meta = {
    description = "Modular Ethereum execution client in Rust by Paradigm";
    homepage = "https://github.com/paradigmxyz/reth";
    license = with lib.licenses; [ mit asl20 ];
    mainProgram = "reth";
    maintainers = with lib.maintainers; [ mitchmindtree ];
```

FliegendeWurst commented on pkgs/by-name/re/reth/package.nix:
```diff
@@ -0,0 +1,60 @@
+{ lib
+, stdenv
```
Please format with the new formatter

mitchmindtree commented:
Thanks for the review @FliegendeWurst !

I've addressed each of your comments.

I've also updated the PR to intialise at the (now stable) version 1.3.12.

I've removed the `Cargo.lock` in favour of using `cargoHash` which uses less space and is friendlier with the `nix-update-script` behaviour.
